ssl/ssl_sess.c

```diff
@@ -1457,4 +1471,193 @@ void SSL_CTX_set_stateless_cookie_verify_cb(
+        return NULL;
+    }
+
+    if ((mode & SSL_SESS_CACHE_NO_INTERNAL_LOOKUP) == 0) {
+        /* Set up the search parameters */
+        memset(&data, 0, sizeof(data));
+        data.cache_id = sc->cache_id;
+        data.cache_id_len = sc->cache_id_len;
+        data.sid_ctx_length = sc->sid_ctx_length;
+        if (sc->sid_ctx_length > 0)
+            memcpy(data.sid_ctx, sc->sid_ctx, sc->sid_ctx_length);
+
+        /* search */
+        if (!CRYPTO_THREAD_read_lock(sc->session_ctx->lock)) {
+            return NULL;
+        }
+        ret = lh_SSL_SESSION_retrieve(sc->session_ctx->sessions, &data);
+        if (ret != NULL)
+            SSL_SESSION_up_ref(ret);
+        CRYPTO_THREAD_unlock(sc->session_ctx->lock);
+
+        if (ret == NULL) {
+            ssl_tsan_counter(sc->session_ctx, &sc->session_ctx->stats.sess_miss);
+        } else if (sess_timedout(ossl_time_now(), ret)) {
+            ssl_tsan_counter(sc->session_ctx, &sc->session_ctx->stats.sess_timeout);
+            SSL_CTX_remove_session(sc->session_ctx, ret);
+            SSL_SESSION_free(ret);
+            ret = NULL;
+        } else {
+            ssl_tsan_counter(sc->session_ctx, &sc->session_ctx->stats.sess_hit);
+        }
+    }
+
+    if (ret == NULL && sc->session_ctx->get_session_cb != NULL) {
+        int copy = 1;
+
+        ret = sc->session_ctx->get_session_cb(s, sc->cache_id,
+                                             sc->cache_id_len,
+                                             &copy);
+
+        if (ret != NULL) {
+            ssl_tsan_counter(sc->session_ctx, &sc->session_ctx->stats.sess_hit);
+
+            /*
+             * Increment reference count now if the session callback asks us
+             * to do so (note that if the session structures returned by the
+             * callback are shared between threads, it must handle the
+             * reference count itself [i.e. copy == 0], or things won't be
+             * thread-safe).
+             */
+            if (copy)
+                SSL_SESSION_up_ref(ret);
+
+            /*
+             * Add the externally cached session to the internal cache as
+             * well if and only if we are supposed to.
+             */
+            if ((mode & SSL_SESS_CACHE_NO_INTERNAL_STORE) == 0) {
+                /*
+                 * Either return value of SSL_CTX_add_session should not
+                 * interrupt the session resumption process. The return
+                 * value is intentionally ignored.
+                 */
+                SSL_CTX_add_session(sc->session_ctx, ret);
+            }
+        }
+    }
+
+    return ret;
+}
+
+long SSL_SESSION_set_timeout_update_cache(const SSL *s, time_t t)
+{
+    SSL_SESSION *ss, *ret;
+    SSL_CONNECTION *sc = SSL_CONNECTION_FROM_SSL(s);
+    long r = 1;
+
+    if (sc == NULL)
+        return 0;
+
+    ss = sc->session;
+    if (ss == NULL)
+        return 0;
+
+    if (!SSL_SESSION_set_timeout(ss, (long)t))
```

Comment:
This mutates an extant session, I suspect this function should not be implemented. Any timeout should be set along with the cache id, before the start of the handshake. If the cache id is set first and automatically loads a session, the new timeout should be ignored, it is only for "fresh" sessions.